Src/VimCore/EditorUtil.fs

```diff
@@ -254,49 +296,140 @@ type SnapshotLineRange  =
 /// This is the representation of a point within a particular line.  It's common
 /// to represent a column in vim and using a SnapshotPoint isn't always the best
 /// representation.  Finding the containing ITextSnapshotLine for a given 
-/// SnapshotPoint is an allocating operation and often shows up as a critical 
-/// metric in profiling.  This structure pairs the two together in a type safe fashion
+/// SnapshotPoint is an allocating operation and often shows up as a critical
+/// metric in profiling.  This structure pairs the two together in a type safe fashion.
+///
+/// Conceptually, a "column" corresponds to a single logical character:

```

Comment:
I'm concerned about merging the functionality here. I think of SnapshotColumn as a core primitive like SnapshotPoint. If there is a version for surrogate pairs I would expect that to be a separate type as well. One which pretty easily converts back and forth.

Reply:
My logic behind adapting `SnapshotColumn` is that every single caller is already expecting to operate on logical characters. If we were to create a new type, there would be no callers using `SnapshotColumn`.

To put it another way, any user of the old `SnapshotColumn` needs to handle surrogate pairs itself, which is 1) hard, and 2) messy (there can be many surrogate pairs on a line), and 3) inefficient. It's inefficient because to know the logical columns on a line you need to parse the text of the whole line. If we ever have to do that, it's better to do it once when we create the column.

This is the beauty of `SnapshotColumn` : VsVim already has it as an efficient abstraction over the raw text. By adapting `SnapshotColumn` to allow variable-width columns, all of VsVim suddenly becomes robust for documents containing surrogate pairs.

Reply:
I don't know that I will convince you, and if not, you can at least use all the tests.

But, I will try one more line of defense.  How many times in VsVim have you used `SnapshotPointUtil.AddOne` or `SubtractOne` and ended up in the middle of a linebreak and had to special case the code? The middle of a line break and the middle of a surrogate pair are traps waiting to catch you. From vim's point of view (and from Visual Studio's point of view) both a line break and a surrogate pair are a single logical character. Visual Studio won't even let you place the caret at those locations. Like it or not, the presence of surrogate pairs means that a `TextBuffer` is effectively encoded. Unless we have an abstraction that allows us to decode the buffer and operate efficiently on logical characters, the alternative is a multitude of unnecessary and probably incomplete special cases.

Reply:
I will take another crack at this basically leaving `SnapshotColumn` untouched.

Reply:
I restored `SnapshotColumn` and created a new type `SnapshotCharacterSpan`. If you don't like that name, I can rename it.